### ISSUE TITLE
Compile Error on CuDnnConvolutionEngine.cu in Ubuntu

### ISSUE THREAD

**Issue by funguy74** [Microsoft/CNTK#271]
Hello

I'm stuck with unknown compile error on CuDnnConvolutionEngine.cu

First, I installed all of required libraries:

$$ ../../configure
Defaulting to --with-buildtype=release
Found cuda at /usr/local/cuda-7.0
Found gdk at /usr/.
Found CUB at /usr/local/cub-1.4.1
Found cuDNN at /usr/local/cudnn-4.0
Found OpenCV at /usr
Cannot locate libzip files
ImageReader will be built without zip container support.
Generating /home/jaey1.kim/CNTK/build/release/Config.make
Generating /home/jaey1.kim/CNTK/build/release/Makefile
run
>make -j all
to build

When I did 'make -j all', I got following errors:

mpic++ -c Source/Math/cudalib.cpp -o /home/jaey1.kim/CNTK/build/release/.build/Source/Math/cudalib.o -D_POSIX_SOURCE -D_XOPEN_SOURCE=600 -D__USE_XOPEN2K -std=c++11 -DUSE_CUDNN -DUSE_ACML -DNDEBUG -DNO_SYNC  -msse3 -std=c++0x -fopenmp -fpermissive -fPIC -Werror -fcheck-new -Wno-error=literal-suffix -g -O4 -ISource/Common/Include -ISource/Math -ISource/CNTK -ISource/ActionsLib -ISource/ComputationNetworkLib -ISource/SGDLib -ISource/SequenceTrainingLib -ISource/CNTK/BrainScript -ISource/Readers/ReaderLib -I/usr/./include/nvidia/gdk -I/usr/local/cub-1.4.1 -I/usr/local/cuda-7.0/include -I/usr/local/cudnn-4.0/cuda/include -I/opt/acml5.3.1/ifort64_mp/include -I/usr/include -MD -MP -MF /home/jaey1.kim/CNTK/build/release/.build/Source/Math/cudalib.d
bin-placing deployable resource files
cp -f Source/CNTK/BrainScript/CNTKCoreLib/CNTK.core.bs /home/jaey1.kim/CNTK/build/release/bin/cntk.core.bs
Source/Math/CuDnnConvolutionEngine.cu(128): error: identifier "cudnnSetFilter4dDescriptor_v4" is undefined

Source/Math/CuDnnConvolutionEngine.cu(128): error: no instance of function template "CudaCall" matches the argument list
            argument types are: (<error-type>, const char [149], const char [6], cudnnStatus_t)

Source/Math/CuDnnConvolutionEngine.cu(518): error: identifier "cudnnConvolutionFwdAlgoPerf_t" is undefined

Source/Math/CuDnnConvolutionEngine.cu(519): error: identifier "cudnnConvolutionBwdDataAlgoPerf_t" is undefined

Source/Math/CuDnnConvolutionEngine.cu(520): error: identifier "cudnnConvolutionBwdFilterAlgoPerf_t" is undefined

Source/CNTK/CNTK.cpp:43:23: fatal error: buildinfo.h: No such file or directory
 #include "buildinfo.h"
                       ^
compilation terminated.
Source/Math/CuDnnConvolutionEngine.cu(485): error: name followed by "::" must be a class or namespace name
          detected during:
            instantiation of class "Microsoft::MSR::CNTK::CuDnnConvolutionEngine<ElemType>::ConvAlgoInfo<T> [with ElemType=float, T=<error-type>]"
(518): here
            instantiation of class "Microsoft::MSR::CNTK::CuDnnConvolutionEngine<ElemType> [with ElemType=float]"
Source/Common/Include/Platform.h(255): here
            instantiation of "std::unique_ptr<T, std::default_delete<T>> std::make_unique<T,Args...>(Args &&...) [with T=Microsoft::MSR::CNTK::CuDnnConvolutionEngine<float>, Args=<int &, Microsoft::MSR::CNTK::ImageLayoutKind &, size_t &, Microsoft::MSR::CNTK::BatchNormImpl &>]"
(635): here
.
.
.

It looks like library compatibility issue but I'm not sure how to resolve this. It would be great if anyone can give me any feedback on this.

**Comment by Alexey-Kamenev** [comment]
Most likely you have an older version of cuDNN in CUDA Toolkit itself (I think TensorFlow or some other toolkits require that for some reason).

**Comment by funguy74** [comment]
Thanks for your suggestion. I looked into the header file and turns out that requested templates are missing such as cudnnSetFilter4dDescriptor_v4. I might need to reinstall cuda I guess.

**Comment by Alexey-Kamenev** [comment]
You don't have to re-install CUDA Toolkit itself but you will need to install cuDNN v4 and make sure you either follow instructions on our Wiki (and remove your current cuDNN .h/.so from CUDA Tookit folders) or install cuDNN v4 directly into CUDA directories so v4 files will overwrite whatever version you currently have. Note that the latter approach may break other toolkits which expect a previous version of cuDNN located in CUDA Toolkit directories.

**Comment by funguy74** [comment]
Thanks for your comments. I can now compile by update into  CuDNN v4. :)
